---
title: "Using `footprintR`"
output:
    rmarkdown::html_document:
        toc: true
        toc_float: true
vignette: >
  %\VignetteIndexEntry{Using `footprintR`}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Overview

`footprintR` provides tools for working with single-molecule footprinting data
in R. These include functions for reading collapsed or read-level data from
files generated by [modkit](https://nanoporetech.github.io/modkit) and
representation of such data as R objects, functions to manipulate such objects
and for visualization..

Current contributors include:

- [Panagiotis Papasaikas](https://github.com/ppapasaikas)
- [Sebastien Smallwood](https://github.com/smalseba)
- [Charlotte Soneson](https://github.com/csoneson)
- [Michael Stadler](https://github.com/mbstadler)

## Installation

`footprintR` can be installed from GitHub via the 
`BiocManager` package:

```
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("fmicompbio/footprintR")
```

## Functionality

Here is a minimal example for using `footprintR` that illustrates its
functionality. We start by loading the package:

```{r load-package, message=FALSE}
library(footprintR)
library(SummarizedExperiment)
```

### Read data

The example is using small experimental datasets that are provided as part of
the package. Here we get the file names for this data:

```{r data-files}
# collapsed 5mC data for a small genomic window in BedMethyl format
bedmethylfiles <- system.file("extdata",
                              c("modkit_pileup_1.bed.gz",
                                "modkit_pileup_2.bed.gz"),
                              package = "footprintR")

# file with sequence of the reference genome in fasta format
reffile <- system.file("extdata", "reference.fa.gz", package = "footprintR")
```

This allows us to read the data using `readBedMethyl()`:

```{r read-data}
names(bedmethylfiles) <- c("sample1", "sample2")
se <- readBedMethyl(bedmethylfiles,
                    sequence.context = 3,
                    sequence.reference = reffile)
se
```
This kind of collapsed data (where individual reads have been combined per
genomic position) is read into a `RangedSummarizedExperiment` object. The
rows here correspond to genomic positions:

```{r rowranges}
rowRanges(se)
```
... and the columns correspond to the different samples (here corresponding
to the names of the two input files in `bedmethylfiles`):

```{r colnames}
colnames(se)
```

The data is stored in two matrices (assays) called `Nmod` (the number of
modified bases per sample and position) and `Nvalid` (the number of valid
total read covering that position in each sample):

```{r assays}
assayNames(se)

head(assay(se, "Nmod"))
head(assay(se, "Nvalid"))
```
From these, you can easily calculate the fraction of modifications:

```{r fraction-modified}
fraction_modified <- assay(se, "Nmod") / assay(se, "Nvalid")
head(fraction_modified)
```
Please note that non-finite fractions result from positions that had
zero coverage in a given sample.

### Plot data

Use the `plotRegion()` function to visualize data for a specific region,
here visualizing the accessibility using GpC methylation (sequence context GCH),
and adding a smooth line:

```{r plot-region-GCH, warning=FALSE, fig.width=6, fig.height=4}
plotRegion(se,
           region = "chr1:6939000-6946000",
           tracks.summary = list(FracMod = "PointSmooth"),
           sequence.context = "GCH")
```

Or the endogenous DNA methylation (sequence context HCG):

```{r plot-region-HCG, warning=FALSE, fig.width=6, fig.height=4}
plotRegion(se,
           region = "chr1:6939000-6946000",
           tracks.summary = list(FracMod = "PointSmooth"),
           sequence.context = "HCG")
```

Or just plotting the coverage (`Nvalid`) as a smooth line:

```{r plot-region-Nvalid, warning=FALSE, fig.width=6, fig.height=4}
plotRegion(se,
           region = "chr1:6939000-6946000",
           tracks.summary = list(Nvalid = "Smooth"))
```

## Session info

```{r session-info}
sessionInfo()
```

