<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Add a variety of read-level modified basecalling summary statistics to a SummarizedExperiment — addReadStats • footprintR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Add a variety of read-level modified basecalling summary statistics to a SummarizedExperiment — addReadStats"><meta name="description" content="This function calculates various per-read summary statistics on modification
probabilities or calls from a SummarizedExperiment
object with genomic positions in rows and reads in columns and adds them to the
colData.
See details for more information on the statistics that are calculated."><meta property="og:description" content="This function calculates various per-read summary statistics on modification
probabilities or calls from a SummarizedExperiment
object with genomic positions in rows and reads in columns and adds them to the
colData.
See details for more information on the statistics that are calculated."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">footprintR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/read-level-data.html">Exploring read-level footprinting data with `footprintR`</a></li>
    <li><a class="dropdown-item" href="../articles/using-footprintR.html">Using `footprintR`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/footprintR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Add a variety of read-level modified basecalling summary statistics to a SummarizedExperiment</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/footprintR/blob/devel/R/addReadStats.R" class="external-link"><code>R/addReadStats.R</code></a></small>
      <div class="d-none name"><code>addReadStats.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates various per-read summary statistics on modification
probabilities or calls from a <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code>
object with genomic positions in rows and reads in columns and adds them to the
<code>colData</code>.
See details for more information on the statistics that are calculated.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">addReadStats</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  regions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sequence.context <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  stats <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min.Nobs.ppos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min.Nobs.pread <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  LowConf <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  LagRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">12</span>, <span class="fl">64</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-se">se<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">RangedSummarizedExperiment</a></code>
object with assay <code>"mod_prob"</code> typically returned by
<code><a href="readModkitExtract.html">readModkitExtract</a></code> or <code><a href="readModBam.html">readModBam</a></code>.</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> object limiting the
positions included in the calculations to the ones overlapping the
corresponding genomic regions. Alternatively, regions can be
specified as a character vector (e.g. "chr1:1200-1300") that can be
coerced into a <code>GRanges</code> object.</p></dd>


<dt id="arg-sequence-context">sequence.context<a class="anchor" aria-label="anchor" href="#arg-sequence-context"></a></dt>
<dd><p>A character vector with sequence context(s)
to include in the calculations. Only positions that match one of the
provided sequence contexts will be included. Sequence contexts can be
provided using IUPAC redundancy codes. The sequence contexts of modified
bases are obtained from <code>rowData(se)$sequence.context</code> and thus
requires that <code>se</code> contains the appropriate information, for example
by setting the <code>sequence.context</code> and <code>sequence.reference</code>
arguments of <code><a href="readModkitExtract.html">readModkitExtract</a></code> when it was generated,
or by adding it using <code><a href="addSeqContext.html">addSeqContext</a></code>.</p></dd>


<dt id="arg-stats">stats<a class="anchor" aria-label="anchor" href="#arg-stats"></a></dt>
<dd><p>Character vector specifying which statistics to calculate.
When set to <code>NULL</code> all available statistics are calculated. See
details for a complete list of available read statistics.</p></dd>


<dt id="arg-min-nobs-ppos">min.Nobs.ppos<a class="anchor" aria-label="anchor" href="#arg-min-nobs-ppos"></a></dt>
<dd><p>A numeric scalar value &gt;=1 indicating the minimum
coverage on individual positions for them to be included in the
calculations. In high coverage data this is an effective filter for
removing spurious modbases, typically the result of erroneous
basecalling. The default <code>NULL</code> sets its value to Q3-0.5*IQR, where
Q3 and IQR are the third quartile and interquartile range of the coverage
distribution estimated from the data in <code>se</code>.</p></dd>


<dt id="arg-min-nobs-pread">min.Nobs.pread<a class="anchor" aria-label="anchor" href="#arg-min-nobs-pread"></a></dt>
<dd><p>A numeric scalar with the minimum number of observed
modifiable bases per read for it to be included in the calculations.
<code>NA</code> values are returned for the reads that do not pass this
threshold.</p></dd>


<dt id="arg-lowconf">LowConf<a class="anchor" aria-label="anchor" href="#arg-lowconf"></a></dt>
<dd><p>A numeric scalar with the minimum call confidence below which
calls are considered "low confidence".</p></dd>


<dt id="arg-lagrange">LagRange<a class="anchor" aria-label="anchor" href="#arg-lagrange"></a></dt>
<dd><p>A numeric vector of two values (minimum and maxium) defining
the range of lags for the calculation of autocorrelation and partial
autocorrelation (see details section).</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>If <code>TRUE</code>, report on progress.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> object
with <code>colData</code> filtered for positions according to <code>regions</code>,
<code>sequence.context</code> and <code>min.Nobs.ppos</code> arguments and extended
to include the read statistics in its row- and column-data.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Calculates a collection of location/scatter statistics and information
theoretic/signal-processing metrics for the modification probability,
confidence or modification call value vectors across individual reads. When
<code>sequence.context</code>, <code>min.coverage</code> or <code>min.Nobs.ppos</code> filters
are enforced, only modifiable bases passing the filters are included in the
calculations. The following statistics are available:</p><dl><dt>MeanModProb</dt>
<dd><p>: Mean modification probability across the read.</p></dd>

<dt>FracMod</dt>
<dd><p>: Fraction of confidently modified bases, defined as the
ratio of modifiable bases with modification probability &gt;= 0.5 over
all modifiable bases.</p></dd>

<dt>MeanConf</dt>
<dd><p>:  Mean call confidence across the read.</p></dd>

<dt>MeanConfUnm</dt>
<dd><p>: Mean call confidence confined to unmodified bases
(modifiable bases with modification probability &lt; 0.5).</p></dd>

<dt>MeanConfMod</dt>
<dd><p>: Mean call confidence confined to modified bases
(modifiable bases with modification probability &gt;= 0.5).</p></dd>

<dt>FracLowConf</dt>
<dd><p>: Fraction of modifiable bases called with low
confidence (call confidence &lt; <code>LowConf</code>).</p></dd>

<dt>IQRModProb</dt>
<dd><p>: Interquartile range of modification probabilities
across the read.</p></dd>

<dt>sdModProb</dt>
<dd><p>: Standard deviation of modification probabilities
across the read.</p></dd>

<dt>SEntrModProb</dt>
<dd><p>: Sample entropy of the modification probability
signal. Sample entropy is a metric assessing the complexity of
one-dimensional physiological signals. The higher the sample entropy,
the more irregular, unpredictable and therefore complex the signal.
See <a href="https://en.wikipedia.org/wiki/Sample_entropy" class="external-link">wikipedia:Sample_entropy</a>
for more details.</p></dd>

<dt>Lag1DModProb</dt>
<dd><p>: Mean Lag1 differences of modification calls,
defined as: <code>mean(Mod[i]-Mod[i-1])</code>, where <code>Mod</code> is a
<code>{0,1}</code> modification call.</p></dd>

<dt>ACModProb</dt>
<dd><p>: Autocorrelation of the modification probability values
for lags in the range <code>LagRange</code>. This range typically covers
the signal of nucleosome periodicity.</p></dd>

<dt>PACModProb</dt>
<dd><p>: Partial autocorrelation of the modification
probability values for lags in the range <code>LagRange</code>. This range
typically covers the signal of nucleosome periodicity.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Panagiotis Papapasaikas</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pmin.int, rank, rbind, rownames, sapply, setdiff, table, tapply,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     union, unique, unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:utils’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     findMatches</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘Biobase’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:MatrixGenerics’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-in"><span><span class="va">modbamfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"6mA_1_10reads.bam"</span>,</span></span>
<span class="r-in"><span>                          package <span class="op">=</span> <span class="st">"footprintR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="readModBam.html">readModBam</a></span><span class="op">(</span>bamfile <span class="op">=</span> <span class="va">modbamfile</span>, regions <span class="op">=</span> <span class="st">"chr1:6940000-6955000"</span>,</span></span>
<span class="r-in"><span>           modbase <span class="op">=</span> <span class="st">"a"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> extracting base modifications from modBAM files</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     opening input file /Users/runner/work/_temp/Library/footprintR/extdata/6mA_1_10reads.bam</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     reading alignments overlapping any of 1 regions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     removed 1724 unaligned (e.g. soft-masked) of 25090 called bases</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     read 3 alignments</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> finding unique genomic positions...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> collapsed 11300 positions to 4772 unique ones</span>
<span class="r-in"><span><span class="va">se_withReadStats</span> <span class="op">&lt;-</span> <span class="fu">addReadStats</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se_withReadStats</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 4772 rows and 0 columns</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se_withReadStats</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 5 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         sample     modbase   n_reads                  qscore</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    &lt;character&gt; &lt;character&gt; &lt;integer&gt;                  &lt;List&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> s1          s1           a         3 14.1428,16.0127,20.3082</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                                                        QC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                                                    &lt;List&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> s1 0.1813986:0.1540698:0.931635:...,0.0969716:0.0632411:0.944381:...,0.0714111:0.0353929:0.950393:...,...</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Stadler, Charlotte Soneson, Panagiotis Papasaikas, Sebastien Smallwood.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

    </footer></div>





  </body></html>

