<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate distances between modified bases on individual reads. — calcModbaseSpacing • footprintR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate distances between modified bases on individual reads. — calcModbaseSpacing"><meta name="description" content="Calculate the frequencies of same read modified base distances,
for example from read-level modification data to estimate nucleosome
repeat length. Distance calculations are implemented in C++
(calcAndCountDist) for efficiency."><meta property="og:description" content="Calculate the frequencies of same read modified base distances,
for example from read-level modification data to estimate nucleosome
repeat length. Distance calculations are implemented in C++
(calcAndCountDist) for efficiency."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">footprintR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/read-level-data.html">Exploring read-level footprinting data with `footprintR`</a></li>
    <li><a class="dropdown-item" href="../articles/using-footprintR.html">Using `footprintR`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/footprintR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate distances between modified bases on individual reads.</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/footprintR/blob/devel/R/modbase_spacing.R" class="external-link"><code>R/modbase_spacing.R</code></a></small>
      <div class="d-none name"><code>calcModbaseSpacing.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Calculate the frequencies of same read modified base distances,
for example from read-level modification data to estimate nucleosome
repeat length. Distance calculations are implemented in C++
(<code><a href="calcAndCountDist.html">calcAndCountDist</a></code>) for efficiency.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calcModbaseSpacing</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  assay.type <span class="op">=</span> <span class="st">"mod_prob"</span>,</span>
<span>  min_mod_prob <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  rmdup <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dmax <span class="op">=</span> <span class="fl">1000L</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-se">se<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p><code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> object
with read-level footprinting data, for example returned by
<code><a href="readModBam.html">readModBam</a></code>. Rows should correspond to positions
and columns to samples.</p></dd>


<dt id="arg-assay-type">assay.type<a class="anchor" aria-label="anchor" href="#arg-assay-type"></a></dt>
<dd><p>A string or integer scalar specifying the assay of
<code>se</code> containing the read-level modification probabilities.</p></dd>


<dt id="arg-min-mod-prob">min_mod_prob<a class="anchor" aria-label="anchor" href="#arg-min-mod-prob"></a></dt>
<dd><p>Numeric scalar giving the minimal modification
probability for a modified base.</p></dd>


<dt id="arg-rmdup">rmdup<a class="anchor" aria-label="anchor" href="#arg-rmdup"></a></dt>
<dd><p>Logical scalar indicating if duplicates should be removed.
If <code>TRUE</code> (the default), only one of several alignments starting at
the same coordinate is used.</p></dd>


<dt id="arg-dmax">dmax<a class="anchor" aria-label="anchor" href="#arg-dmax"></a></dt>
<dd><p>Numeric scalar specifying the maximal distance between
modified bases on the same read to count.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>integer</code> vector with <code>dmax</code> elements, with the element at
position <code>d</code> giving the observed number of alignment pairs at that
distance.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Phasograms were originally described in Valouev et al., Nature
2011 (doi:10.1038/nature10002). The implementation here differs in three
ways from the original algorithms:</p><ol><li><p>Instead of same strand alignment start positions, this function
is adapted to single-molecule footprinting data and measures
the distances between same-read modified base positions.</p></li>
<li><p>It does not implement removing of positions that have been seen
less than <code>n</code> times (referred to as a <code>n</code>-pile subset in
the paper).</p></li>
<li><p>It does allow to retain only alignments that fall into selected
genomic intervals (<code>regions</code> argument).</p></li>
</ol></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="estimateNRL.html">estimateNRL</a></code> to estimate the nucleosome repeat length
from a phasogram, <code><a href="plotModbaseSpacing.html">plotModbaseSpacing</a></code> to visualize an annotated
phasogram, <code><a href="calcAndCountDist.html">calcAndCountDist</a></code> for low-level distance counting.</p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Michael Stadler</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">modbamfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>                           <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"6mA_1_10reads.bam"</span>, <span class="st">"6mA_2_10reads.bam"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                           package <span class="op">=</span> <span class="st">"footprintR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="readModBam.html">readModBam</a></span><span class="op">(</span><span class="va">modbamfiles</span>, <span class="st">"chr1:6940000-6955000"</span>, <span class="st">"a"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">moddist</span> <span class="op">&lt;-</span> <span class="fu">calcModbaseSpacing</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="estimateNRL.html">estimateNRL</a></span><span class="op">(</span><span class="va">moddist</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $nrl</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 183.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $nrl.CI95</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2.5 %   97.5 % </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 179.8787 187.9213 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="plotModbaseSpacing.html">plotModbaseSpacing</a></span><span class="op">(</span><span class="va">moddist</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calcModbaseSpacing-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="plotModbaseSpacing.html">plotModbaseSpacing</a></span><span class="op">(</span><span class="va">moddist</span>, detailedPlots <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calcModbaseSpacing-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Stadler, Charlotte Soneson, Panagiotis Papasaikas, Sebastien Smallwood.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

    </footer></div>





  </body></html>

