<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Call modkit extract to extract read-level base modifications from modBAM files — modkitExtract • footprintR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Call modkit extract to extract read-level base modifications from modBAM files — modkitExtract"><meta name="description" content="This function is a wrapper around the modkit extract sub-command to extract
read-level base modification information from modBAM files into tab-separated
values table(s).
For more information on available modkit extract arguments and output
tables specification see https://nanoporetech.github.io/modkit/intro_extract.html"><meta property="og:description" content="This function is a wrapper around the modkit extract sub-command to extract
read-level base modification information from modBAM files into tab-separated
values table(s).
For more information on available modkit extract arguments and output
tables specification see https://nanoporetech.github.io/modkit/intro_extract.html"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">footprintR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/read-level-data.html">Exploring read-level footprinting data with `footprintR`</a></li>
    <li><a class="dropdown-item" href="../articles/using-footprintR.html">Using `footprintR`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/footprintR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Call <code>modkit extract</code> to extract read-level base modifications from modBAM files</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/footprintR/blob/devel/R/modkitExtract.R" class="external-link"><code>R/modkitExtract.R</code></a></small>
      <div class="d-none name"><code>modkitExtract.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is a wrapper around the <code>modkit extract</code> sub-command to extract
read-level base modification information from modBAM files into tab-separated
values table(s).
For more information on available <code>modkit extract</code> arguments and output
tables specification see https://nanoporetech.github.io/modkit/intro_extract.html</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">modkitExtract</span><span class="op">(</span></span>
<span>  modkit_bin <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">bamfile</span>,</span>
<span>  regions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  num_reads <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  out_extract_table <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  out_read_calls <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  out_log_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  modkit_args <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tempdir_base <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-modkit-bin">modkit_bin<a class="anchor" aria-label="anchor" href="#arg-modkit-bin"></a></dt>
<dd><p>Character scalar specifying the path to the <code>modkit</code>
binary. If <code>NULL</code>, <code>modkit</code> will be searched on the path using
<code><a href="https://rdrr.io/r/base/Sys.which.html" class="external-link">Sys.which</a></code>.</p></dd>


<dt id="arg-bamfile">bamfile<a class="anchor" aria-label="anchor" href="#arg-bamfile"></a></dt>
<dd><p>Character scalar specifying the path to a <code>modBAM</code>
file. An indexed BAM file can significantly speed up certain extract
operations.</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> object specifying which
genomic regions to extract the reads from. Note that the reads are not
trimmed to the boundaries of the specified ranges. As a result, returned
positions will typically extend out of the specified regions.</p></dd>


<dt id="arg-num-reads">num_reads<a class="anchor" aria-label="anchor" href="#arg-num-reads"></a></dt>
<dd><p>Number of reads to extract per specified genomic region.
When <code>N</code> genomic ranges are specified, the total number of extracted
reads will be at most <code>num_reads * N</code>.</p></dd>


<dt id="arg-out-extract-table">out_extract_table<a class="anchor" aria-label="anchor" href="#arg-out-extract-table"></a></dt>
<dd><p>Character scalar specifying the path for the
<code>extract table</code> output. Can be <code>NULL</code> if only a <code>read-calls</code>
table is needed.</p></dd>


<dt id="arg-out-read-calls">out_read_calls<a class="anchor" aria-label="anchor" href="#arg-out-read-calls"></a></dt>
<dd><p>Character scalar specifying the path for the
<code>read-calls table</code> output. Can be <code>NULL</code> if only an <code>extract table</code>
is needed.</p></dd>


<dt id="arg-out-log-file">out_log_file<a class="anchor" aria-label="anchor" href="#arg-out-log-file"></a></dt>
<dd><p>Character scalar specifying the path for the the command
run log output. Can be <code>NULL</code> if no log file is needed (not
recommended).</p></dd>


<dt id="arg-modkit-args">modkit_args<a class="anchor" aria-label="anchor" href="#arg-modkit-args"></a></dt>
<dd><p>Character vector with additional <code>modkit extract</code>
arguments. Please refer to the <code>modkit extract</code> documentation for a
complete list of possible arguments.</p></dd>


<dt id="arg-tempdir-base">tempdir_base<a class="anchor" aria-label="anchor" href="#arg-tempdir-base"></a></dt>
<dd><p>Character scalar specifying the path to create the
<code>modkit_temp</code> directory. This temporary directory is only created when
multiple genomic regions are specified.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>A logical scalar. If <code>TRUE</code>, report on progress.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A character vector with the elements "extract-table", "read-calls"
and "run-log", specifying the paths to the generated table, call and log
files, respectively. An <code>NA</code> value indicates that a given output
was not generated.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="https://nanoporetech.github.io/modkit" class="external-link"><code>modkit</code> software</a>,
<a href="https://nanoporetech.github.io/modkit/intro_extract.html" class="external-link"><code>modkit extract</code> documentation and tabulated output formats specification</a>,
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> for the object used to specify
genomic regions.</p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Panagiotis Papapasaikas</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">GR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:12340678-12345678"</span>, <span class="st">"chr2:12340678-12345678"</span><span class="op">)</span>, <span class="st">"GRanges"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">further_args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>   <span class="st">'-t 8'</span>, <span class="st">'--force'</span>,</span></span>
<span class="r-in"><span>   <span class="st">'--mapped-only'</span>, <span class="st">'--edge-filter 100'</span>,</span></span>
<span class="r-in"><span>   <span class="st">'-p  0.2'</span>,</span></span>
<span class="r-in"><span>   <span class="st">'--mod-threshold m:0.8'</span>,</span></span>
<span class="r-in"><span>   <span class="st">'--mod-threshold a:0.5'</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Produce both `extract table` and `read calls` files for multiple GRanges</span></span></span>
<span class="r-in"><span><span class="fu">modkitExtract</span><span class="op">(</span>modkit <span class="op">=</span> <span class="va">modkit_bin_PATH</span>, bamfile <span class="op">=</span> <span class="va">BAMF</span>, num_reads <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>              regions <span class="op">=</span> <span class="va">GR</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, out_extract_table <span class="op">=</span> <span class="st">"test.etbl"</span>,</span></span>
<span class="r-in"><span>              out_read_calls <span class="op">=</span> <span class="st">"test.rdcl"</span>, modkit_args <span class="op">=</span> <span class="va">further_args</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Produce only `extract table` file</span></span></span>
<span class="r-in"><span><span class="fu">modkitExtract</span><span class="op">(</span>modkit <span class="op">=</span> <span class="va">modkit_bin_PATH</span>, bamfile <span class="op">=</span> <span class="va">BAMF</span>, num_reads <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>              regions <span class="op">=</span> <span class="va">GR</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, out_extract_table <span class="op">=</span> <span class="st">"test.etbl"</span>,</span></span>
<span class="r-in"><span>              out_read_calls <span class="op">=</span> <span class="cn">NULL</span>, modkit_args <span class="op">=</span> <span class="va">further_args</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Produce only `read calls` file</span></span></span>
<span class="r-in"><span><span class="fu">modkitExtract</span><span class="op">(</span>modkit <span class="op">=</span> <span class="va">modkit_bin_PATH</span>, bamfile <span class="op">=</span> <span class="va">BAMF</span>, num_reads <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>              regions <span class="op">=</span> <span class="va">GR</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, out_extract_table <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>              out_read_calls <span class="op">=</span> <span class="st">"test.rdcl"</span>, modkit_args <span class="op">=</span> <span class="va">further_args</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Stadler, Charlotte Soneson, Panagiotis Papasaikas, Sebastien Smallwood.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

    </footer></div>





  </body></html>

