<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Read modkit extract file(s) — readModkitExtract • footprintR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Read modkit extract file(s) — readModkitExtract"><meta name="description" content="This function reads one or more modkit extract files and generates a
single SummarizedExperiment object with rows representing positions and
columns representing samples. The SummarizedExperiment object
contains one assay (mod_prob) with modification probabilities for
each position in each read. Unobserved read/position combinations are
represented by a zero, while all values that are 'implicitly' called by
modkit (with a modification probability less than 5%) are represented with
a value of 0.02."><meta property="og:description" content="This function reads one or more modkit extract files and generates a
single SummarizedExperiment object with rows representing positions and
columns representing samples. The SummarizedExperiment object
contains one assay (mod_prob) with modification probabilities for
each position in each read. Unobserved read/position combinations are
represented by a zero, while all values that are 'implicitly' called by
modkit (with a modification probability less than 5%) are represented with
a value of 0.02."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">footprintR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/read-level-data.html">Exploring read-level footprinting data with `footprintR`</a></li>
    <li><a class="dropdown-item" href="../articles/using-footprintR.html">Using `footprintR`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/footprintR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Read modkit extract file(s)</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/footprintR/blob/devel/R/readModkitExtract.R" class="external-link"><code>R/readModkitExtract.R</code></a></small>
      <div class="d-none name"><code>readModkitExtract.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function reads one or more modkit extract files and generates a
single SummarizedExperiment object with rows representing positions and
columns representing samples. The SummarizedExperiment object
contains one assay (<code>mod_prob</code>) with modification probabilities for
each position in each read. Unobserved read/position combinations are
represented by a zero, while all values that are 'implicitly' called by
modkit (with a modification probability less than 5%) are represented with
a value of 0.02.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">readModkitExtract</span><span class="op">(</span></span>
<span>  <span class="va">fnames</span>,</span>
<span>  <span class="va">modbase</span>,</span>
<span>  filter <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nrows <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  seqinfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncpu <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fnames">fnames<a class="anchor" aria-label="anchor" href="#arg-fnames"></a></dt>
<dd><p>Character vector with one or several paths of <code>extract.tsv</code>
files, such as generated by <code>modkit</code>. If <code>fnames</code> is a named
vector, the names are used as prefixes for the column (read) names in
the returned <code>SummarizedExperiment</code> object, and as column names
for the object itself. Otherwise, the prefixes/names
will be <code>s1</code>, ..., <code>sN</code>, where <code>N</code> is the length of
<code>fnames</code>.</p></dd>


<dt id="arg-modbase">modbase<a class="anchor" aria-label="anchor" href="#arg-modbase"></a></dt>
<dd><p>Character vector defining the modified base for each sample.
If <code>modbase</code> is a named vector, the names should correspond to
the names of <code>fnames</code>. Otherwise, it will be assumed that the
elements are in the same order as the files in <code>fnames</code>. If
<code>modbase</code> has length 1, the same modified base will be used for
all samples.</p></dd>


<dt id="arg-filter">filter<a class="anchor" aria-label="anchor" href="#arg-filter"></a></dt>
<dd><p>Either <code>NULL</code> (no filtering is performed),
<code>"modkit"</code> (only records where the <code>"fail"</code> column in the
extract file is <code>FALSE</code> are retained), or a named vector, providing
filter thresholds (on the call probability) for the
unmodified and modified bases, respectively. The names of the vector
should be <code>"-"</code> and the value(s) in <code>modbase</code>.</p></dd>


<dt id="arg-nrows">nrows<a class="anchor" aria-label="anchor" href="#arg-nrows"></a></dt>
<dd><p>Only read <code>nrows</code> rows of each input file.</p></dd>


<dt id="arg-seqinfo">seqinfo<a class="anchor" aria-label="anchor" href="#arg-seqinfo"></a></dt>
<dd><p><code>NULL</code> or a <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html" class="external-link">Seqinfo</a></code> object
containing information about the set of genomic sequences (chromosomes).
Alternatively, a named numeric vector with genomic sequence names and
lengths. Useful to set the sorting order of sequence names.</p></dd>


<dt id="arg-ncpu">ncpu<a class="anchor" aria-label="anchor" href="#arg-ncpu"></a></dt>
<dd><p>A numeric scalar giving the number of parallel CPU threads to
to use for some of the steps in <code>readModkitExtract()</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>If <code>TRUE</code>, report on progress.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> object
with genomic positions in rows and samples in columns. The assay
<code>"mod_prob"</code> contains per-read modification probabilities,
with each column (sample) corresponding to a position-by-read
<code><a href="https://rdrr.io/pkg/SparseArray/man/SparseArray-class.html" class="external-link">SparseMatrix</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="https://nanoporetech.github.io/modkit" class="external-link"><code>modkit</code> software</a>,
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> for the returned object type,
<code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></code> for the function used to read the input files</p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Charlotte Soneson</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">extrfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"modkit_extract_rc_5mC_1.tsv.gz"</span>,</span></span>
<span class="r-in"><span>                        package <span class="op">=</span> <span class="st">"footprintR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## ... no filtering</span></span></span>
<span class="r-in"><span><span class="fu">readModkitExtract</span><span class="op">(</span><span class="va">extrfile</span>, modbase <span class="op">=</span> <span class="st">"m"</span>, filter <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: RangedSummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 6432 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(2): modkit_threshold filter_threshold</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): mod_prob</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(6432): chr1:6932662:+ chr1:6932664:+ ... chr1:6942543:-</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chr1:6942552:-</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(1): s1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(2): sample modbase</span>
<span class="r-in"><span><span class="co">## ... modkit filtering</span></span></span>
<span class="r-in"><span><span class="fu">readModkitExtract</span><span class="op">(</span><span class="va">extrfile</span>, modbase <span class="op">=</span> <span class="st">"m"</span>, filter <span class="op">=</span> <span class="st">"modkit"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: RangedSummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 5893 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(2): modkit_threshold filter_threshold</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): mod_prob</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(5893): chr1:6932662:+ chr1:6932664:+ ... chr1:6942543:-</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chr1:6942552:-</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(1): s1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(2): sample modbase</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Stadler, Charlotte Soneson, Panagiotis Papasaikas, Sebastien Smallwood.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

    </footer></div>





  </body></html>

