<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Read base modifications from bam file(s). — readModBam • footprintR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Read base modifications from bam file(s). — readModBam"><meta property="og:description" content="Parse ML and MM tags (see https://samtools.github.io/hts-specs/SAMtags.pdf,
section 1.7) and return a SummarizedExperiment
object with information on modified bases."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">footprintR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/read-level-data.html">Exploring read-level footprinting data with `footprintR`</a>
    </li>
    <li>
      <a href="../articles/using-footprintR.html">Using `footprintR`</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/fmicompbio/footprintR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read base modifications from bam file(s).</h1>
    <small class="dont-index">Source: <a href="https://github.com/fmicompbio/footprintR/blob/devel/R/readModBam.R" class="external-link"><code>R/readModBam.R</code></a></small>
    <div class="hidden name"><code>readModBam.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Parse ML and MM tags (see https://samtools.github.io/hts-specs/SAMtags.pdf,
section 1.7) and return a <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code>
object with information on modified bases.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">readModBam</span><span class="op">(</span></span>
<span>  <span class="va">bamfiles</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  <span class="va">modbase</span>,</span>
<span>  seqinfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncpu <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-bamfiles">bamfiles<a class="anchor" aria-label="anchor" href="#arg-bamfiles"></a></dt>
<dd><p>Character vector with one or several paths of <code>modBAM</code>
files, containing information about base modifications in <code>MM</code> and
<code>ML</code> tags. If <code>bamfiles</code> is a named vector, the names are used
as sample names and prefixes for read names. Otherwise, the prefixes will
be <code>s1</code>, ..., <code>sN</code>, where <code>N</code> is the length of
<code>bamfiles</code>. All <code>bamfiles</code> must have an index.</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> object specifying which
genomic regions to extract the reads from. Alternatively, regions can be
specified as a character vector (e.g. "chr1:1200-1300") that can be
coerced into a <code>GRanges</code> object. Note that the reads are not
trimmed to the boundaries of the specified ranges. As a result, returned
positions will typically extend out of the specified regions.</p></dd>


<dt id="arg-modbase">modbase<a class="anchor" aria-label="anchor" href="#arg-modbase"></a></dt>
<dd><p>Character vector defining the modified base for each sample.
If <code>modbase</code> is a named vector, the names should correspond to
the names of <code>bamfiles</code>. Otherwise, it will be assumed that the
elements are in the same order as the files in <code>bamfiles</code>. If
<code>modbase</code> has length 1, the same modified base will be used for
all samples.</p></dd>


<dt id="arg-seqinfo">seqinfo<a class="anchor" aria-label="anchor" href="#arg-seqinfo"></a></dt>
<dd><p><code>NULL</code> or a <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html" class="external-link">Seqinfo</a></code> object
containing information about the set of genomic sequences (chromosomes).
Alternatively, a named numeric vector with genomic sequence names and
lengths. Useful to set the sorting order of sequence names.</p></dd>


<dt id="arg-ncpu">ncpu<a class="anchor" aria-label="anchor" href="#arg-ncpu"></a></dt>
<dd><p>A numeric scalar giving the number of parallel CPU threads to
to use for some of the steps in <code>readModBam</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical scalar. If <code>TRUE</code>, report on progress.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> object
with genomic positions in rows and samples in columns. The assay
<code>"mod_prob"</code> contains per-read modification probabilities,
with each column (sample) corresponding to a position-by-read
<code><a href="https://rdrr.io/pkg/SparseArray/man/SparseArray-class.html" class="external-link">SparseMatrix</a></code>.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>https://samtools.github.io/hts-specs/SAMtags.pdf describing the
SAM ML and MM tags for base modifications.</p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Michael Stadler</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">modbamfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"6mA_1_10reads.bam"</span>,</span></span>
<span class="r-in"><span>                          package <span class="op">=</span> <span class="st">"footprintR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">readModBam</span><span class="op">(</span>bamfiles <span class="op">=</span> <span class="va">modbamfile</span>, regions <span class="op">=</span> <span class="st">"chr1:6940000-6955000"</span>,</span></span>
<span class="r-in"><span>           modbase <span class="op">=</span> <span class="st">"a"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> extracting base modifications from modBAM files</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     opening input file /Users/runner/work/_temp/Library/footprintR/extdata/6mA_1_10reads.bam</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     reading alignments overlapping any of 1 regions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     removed 1728 unaligned (e.g. soft-masked) of 20554 called bases</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     read 3 alignments</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> finding unique genomic positions...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> collapsed 8350 positions to 6057 unique ones</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: RangedSummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 6057 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): mod_prob</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(6057): chr1:6925830:- chr1:6925834:- ... chr1:6941622:-</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chr1:6941631:-</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(1): s1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(4): sample modbase n_reads qscore</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Michael Stadler, Charlotte Soneson, Panagiotis Papasaikas, Sebastien Smallwood.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

      </footer></div>






  </body></html>

