<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Read collapsed single-molecule footprinting data from a bedMethyl file. — readBedMethyl • footprintR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Read collapsed single-molecule footprinting data from a bedMethyl file. — readBedMethyl"><meta name="description" content="This function will read collapsed single-molecule footprinting data
(reads combined per genomic position) from a bedMethyl file."><meta property="og:description" content="This function will read collapsed single-molecule footprinting data
(reads combined per genomic position) from a bedMethyl file."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">footprintR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/read-level-data.html">Exploring read-level footprinting data with `footprintR`</a></li>
    <li><a class="dropdown-item" href="../articles/using-footprintR.html">Using `footprintR`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/footprintR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Read collapsed single-molecule footprinting data from a <code>bedMethyl</code> file.</h1>
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/footprintR/blob/devel/R/readBedMethyl.R" class="external-link"><code>R/readBedMethyl.R</code></a></small>
      <div class="d-none name"><code>readBedMethyl.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function will read collapsed single-molecule footprinting data
(reads combined per genomic position) from a <code>bedMethyl</code> file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">readBedMethyl</span><span class="op">(</span></span>
<span>  <span class="va">fnames</span>,</span>
<span>  modbase <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nrows <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  seqinfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sequence.context.width <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  sequence.reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncpu <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fnames">fnames<a class="anchor" aria-label="anchor" href="#arg-fnames"></a></dt>
<dd><p>Character vector with one or several paths of <code>bedMethyl</code>
files, such as generated by <code>modkit pileup</code>. Each file will be read
separately and become one of the columns in the returned
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> object.
If <code>fnames</code> is a named vector, the names are used as column names in
the returned object. Otherwise, the column names will be <code>s1</code>, ...,
<code>sN</code>, where <code>N</code> is the length of <code>fnames</code>.
If several elements of <code>fnames</code> have identical
names, the data from the corresponding files are summed into a single
column in the returned object.</p></dd>


<dt id="arg-modbase">modbase<a class="anchor" aria-label="anchor" href="#arg-modbase"></a></dt>
<dd><p>Character vector defining the modified base (or bases) to
read. Useful for reading a subset of the data from <code>bedMethyl</code> files
that contain multiple types of modified bases. If <code>NULL</code> (the
default), all rows in the input file are read.</p></dd>


<dt id="arg-nrows">nrows<a class="anchor" aria-label="anchor" href="#arg-nrows"></a></dt>
<dd><p>Only read <code>nrows</code> rows of the input file.</p></dd>


<dt id="arg-seqinfo">seqinfo<a class="anchor" aria-label="anchor" href="#arg-seqinfo"></a></dt>
<dd><p><code>NULL</code> or a <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html" class="external-link">Seqinfo</a></code> object
containing information about the set of genomic sequences (chromosomes).
Alternatively, a named numeric vector with genomic sequence names and
lengths. Useful to set the sorting order of sequence names.</p></dd>


<dt id="arg-sequence-context-width-sequence-reference">sequence.context.width, sequence.reference<a class="anchor" aria-label="anchor" href="#arg-sequence-context-width-sequence-reference"></a></dt>
<dd><p>Define the sequence
context to be extracted around modified bases. By default (
<code>sequence.context.width = 0</code>), no sequence context will be
extracted, otherwise it will be returned in <code>rowData(x)$sequence.context</code>.
See <code><a href="addSeqContext.html">addSeqContext</a></code> for details.</p></dd>


<dt id="arg-ncpu">ncpu<a class="anchor" aria-label="anchor" href="#arg-ncpu"></a></dt>
<dd><p>A numeric scalar giving the number of parallel CPU threads to
to use for some of the steps in <code>readBedMethyl()</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>If <code>TRUE</code>, report on progress.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> object
with genomic positions in rows and samples (the unique names of
<code>fnames</code>) in the columns. If <code>sequence.context.width != 0</code>,
<code>rowData(x)$sequence.context</code> will be a <code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></code>
object with the extracted sequences.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="https://nanoporetech.github.io/modkit" class="external-link"><code>modkit</code> software</a>,
<a href="https://nanoporetech.github.io/modkit/intro_bedmethyl.html#description-of-bedmethyl-output" class="external-link"><code>bedMethyl</code> format description</a>,
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></code> for the returned object type,
<code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></code> for the function used to read the input files,
<code><a href="addSeqContext.html">addSeqContext</a></code> used to add the sequence context.</p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Michael Stadler</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">bmfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"modkit_pileup_1.bed.gz"</span>, package <span class="op">=</span> <span class="st">"footprintR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">readBedMethyl</span><span class="op">(</span><span class="va">bmfile</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: RangedSummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 10000 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(2): Nmod Nvalid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(1): s1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(0):</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Stadler, Charlotte Soneson, Panagiotis Papasaikas, Sebastien Smallwood.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

    </footer></div>





  </body></html>

