<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Exploring read-level footprinting data with `footprintR` • footprintR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Exploring read-level footprinting data with `footprintR`">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">footprintR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/read-level-data.html">Exploring read-level footprinting data with `footprintR`</a></li>
    <li><a class="dropdown-item" href="../articles/using-footprintR.html">Using `footprintR`</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/footprintR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Exploring read-level footprinting data with `footprintR`</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/footprintR/blob/devel/vignettes/read-level-data.Rmd" class="external-link"><code>vignettes/read-level-data.Rmd</code></a></small>
      <div class="d-none name"><code>read-level-data.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>We start the analysis by loading footprinting data at the level of
individual reads (for an example of working with summary-level data, see
<code><a href="../articles/using-footprintR.html">vignette("using-footprintR")</a></code>). Read-level data can be
extracted from <code>modBam</code> files (standard bam file with
modification data stored in <code>MM</code> and <code>ML</code> tags,
see <a href="https://samtools.github.io/hts-specs/SAMtags.pdf" class="external-link">SAMtags.pdf</a>)
using <code><a href="../reference/readModBam.html">readModBam()</a></code>. This is the preferred way described
below.</p>
<p>Alternatively, read-level modification data can also be extracted
from <code>modBam</code> files using <a href="https://nanoporetech.github.io/modkit" class="external-link">modkit</a> (see
<code><a href="../reference/modkitExtract.html">modkitExtract()</a></code> for calling <code>modkit</code> from R and
<code><a href="../reference/readModkitExtract.html">readModkitExtract()</a></code> to read the resulting output file).
This will require a local installation of <code>modkit</code> and may
only be of interest if specific <code>modkit</code> features should be
used that are not available otherwise.</p>
<p>The <code>footprintR</code> package contains small example
<code>modBam</code> files that were generated using the <a href="https://github.com/nanoporetech/dorado" class="external-link">Dorado</a> aligner:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmicompbio/footprintR" class="external-link">footprintR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read-level 6mA data generated by 'dorado'</span></span>
<span><span class="va">modbamfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>                           <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"6mA_1_10reads.bam"</span>, <span class="st">"6mA_2_10reads.bam"</span><span class="op">)</span>,</span>
<span>                            package <span class="op">=</span> <span class="st">"footprintR"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">modbamfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span></span></code></pre></div>
<p>Modification data is read from these files using
<code><a href="../reference/readModBam.html">readModBam()</a></code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readModBam.html">readModBam</a></span><span class="op">(</span>bamfiles <span class="op">=</span> <span class="va">modbamfiles</span>,</span>
<span>                 regions <span class="op">=</span> <span class="st">"chr1:6940000-7000000"</span>,</span>
<span>                 modbase <span class="op">=</span> <span class="st">"a"</span>,</span>
<span>                 verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; extracting base modifications from modBAM files</span></span>
<span><span class="co">#&gt;     opening input file /Users/runner/work/_temp/Library/footprintR/extdata/6mA_1_10reads.bam</span></span>
<span><span class="co">#&gt;     reading alignments overlapping any of 1 regions</span></span>
<span><span class="co">#&gt;     removed 1728 unaligned (e.g. soft-masked) of 20554 called bases</span></span>
<span><span class="co">#&gt;     read 3 alignments</span></span>
<span><span class="co">#&gt;     opening input file /Users/runner/work/_temp/Library/footprintR/extdata/6mA_2_10reads.bam</span></span>
<span><span class="co">#&gt;     reading alignments overlapping any of 1 regions</span></span>
<span><span class="co">#&gt;     removed 320 unaligned (e.g. soft-masked) of 10174 called bases</span></span>
<span><span class="co">#&gt;     read 2 alignments</span></span>
<span><span class="co">#&gt; finding unique genomic positions...collapsed 14789 positions to 9266 unique ones</span></span>
<span><span class="va">se</span></span>
<span><span class="co">#&gt; class: RangedSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 9266 2 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): mod_prob</span></span>
<span><span class="co">#&gt; rownames(9266): chr1:6930410:+ chr1:6930415:+ ... chr1:6941622:-</span></span>
<span><span class="co">#&gt;   chr1:6941631:-</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(2): sample1 sample2</span></span>
<span><span class="co">#&gt; colData names(4): sample modbase n_reads qscore</span></span></code></pre></div>
<p>This will create a <code>RangedSummarizedExperiment</code> object
with positions in rows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rows are positions...</span></span>
<span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; UnstitchedGPos object with 9266 positions and 0 metadata columns:</span></span>
<span><span class="co">#&gt;                  seqnames       pos strand</span></span>
<span><span class="co">#&gt;                     &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;   chr1:6930410:+     chr1   6930410      +</span></span>
<span><span class="co">#&gt;   chr1:6930415:+     chr1   6930415      +</span></span>
<span><span class="co">#&gt;   chr1:6930419:+     chr1   6930419      +</span></span>
<span><span class="co">#&gt;   chr1:6930421:+     chr1   6930421      +</span></span>
<span><span class="co">#&gt;   chr1:6930428:+     chr1   6930428      +</span></span>
<span><span class="co">#&gt;              ...      ...       ...    ...</span></span>
<span><span class="co">#&gt;   chr1:6941611:-     chr1   6941611      -</span></span>
<span><span class="co">#&gt;   chr1:6941614:-     chr1   6941614      -</span></span>
<span><span class="co">#&gt;   chr1:6941620:-     chr1   6941620      -</span></span>
<span><span class="co">#&gt;   chr1:6941622:-     chr1   6941622      -</span></span>
<span><span class="co">#&gt;   chr1:6941631:-     chr1   6941631      -</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>Just like with summary-level data, columns correspond to samples</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ... and columns are samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 4 columns</span></span>
<span><span class="co">#&gt;              sample     modbase   n_reads                  qscore</span></span>
<span><span class="co">#&gt;         &lt;character&gt; &lt;character&gt; &lt;integer&gt;                  &lt;List&gt;</span></span>
<span><span class="co">#&gt; sample1     sample1           a         3 14.1428,16.0127,20.3082</span></span>
<span><span class="co">#&gt; sample2     sample2           a         2        9.67461,13.66480</span></span></code></pre></div>
<p>The sample names are obtained from the input files (here
<code>extractfiles</code>), or if the files are not named will be
automatically assigned (each file corresponding to a separate
sample).</p>
<p>However, as each sample typically contains several reads (see
<code>se$n_reads</code>), the <code>qscore</code> column that stores the
quality scores for individual reads, is not a simple vector, but a list
of vectors, in which scores are grouped by sample. It may be needed to
<code><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist()</a></code> it if a flat vector is needed:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span><span class="op">$</span><span class="va">qscore</span></span>
<span><span class="co">#&gt; List of length 2</span></span>
<span><span class="co">#&gt; names(2): sample1 sample2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">qscore</span><span class="op">)</span></span>
<span><span class="co">#&gt;  sample1  sample1  sample1  sample2  sample2 </span></span>
<span><span class="co">#&gt; 14.14280 16.01270 20.30820  9.67461 13.66480</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="explore-assay-data">Explore assay data<a class="anchor" aria-label="anchor" href="#explore-assay-data"></a>
</h2>
<p>The single assay <code>mod_prob</code> is a <code>DataFrame</code>
with modification probabilities.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "mod_prob"</span></span>
<span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"mod_prob"</span><span class="op">)</span></span>
<span><span class="va">m</span></span>
<span><span class="co">#&gt; DataFrame with 9266 rows and 2 columns</span></span>
<span><span class="co">#&gt;                        sample1        sample2</span></span>
<span><span class="co">#&gt;                 &lt;SparseMatrix&gt; &lt;SparseMatrix&gt;</span></span>
<span><span class="co">#&gt; chr1:6930410:+           0:0:0         0:0.02</span></span>
<span><span class="co">#&gt; chr1:6930415:+           0:0:0         0:0.02</span></span>
<span><span class="co">#&gt; chr1:6930419:+           0:0:0         0:0.02</span></span>
<span><span class="co">#&gt; chr1:6930421:+           0:0:0         0:0.02</span></span>
<span><span class="co">#&gt; chr1:6930428:+           0:0:0         0:0.02</span></span>
<span><span class="co">#&gt; ...                        ...            ...</span></span>
<span><span class="co">#&gt; chr1:6941611:- 0:0:0.052734375            0:0</span></span>
<span><span class="co">#&gt; chr1:6941614:- 0:0:0.130859375            0:0</span></span>
<span><span class="co">#&gt; chr1:6941620:- 0:0:0.068359375            0:0</span></span>
<span><span class="co">#&gt; chr1:6941622:- 0:0:0.060546875            0:0</span></span>
<span><span class="co">#&gt; chr1:6941631:- 0:0:0.056640625            0:0</span></span></code></pre></div>
<p>Again, to store read-level data for variable numbers of reads in each
sample, each column (sample) is not just a simple vector, but a
position-by-read <code>SparseMatrix</code>. If a simple matrix is
needed, in which columns correspond to reads instead of sample, it can
be created using <code><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span><span class="op">$</span><span class="va">sample1</span></span>
<span><span class="co">#&gt; &lt;9266 x 3 SparseMatrix&gt; of type "double" [nzcount=8350 (30%)]:</span></span>
<span><span class="co">#&gt;         sample1-233e48a7-f379-4dcf-9270-958231125563</span></span>
<span><span class="co">#&gt;    [1,]                                            0</span></span>
<span><span class="co">#&gt;    [2,]                                            0</span></span>
<span><span class="co">#&gt;    [3,]                                            0</span></span>
<span><span class="co">#&gt;    [4,]                                            0</span></span>
<span><span class="co">#&gt;    [5,]                                            0</span></span>
<span><span class="co">#&gt;     ...                                            .</span></span>
<span><span class="co">#&gt; [9262,]                                   0.00000000</span></span>
<span><span class="co">#&gt; [9263,]                                   0.00000000</span></span>
<span><span class="co">#&gt; [9264,]                                   0.00000000</span></span>
<span><span class="co">#&gt; [9265,]                                   0.00000000</span></span>
<span><span class="co">#&gt; [9266,]                                   0.00000000</span></span>
<span><span class="co">#&gt;         sample1-d52a5f6a-a60a-4f85-913e-eada84bfbfb9</span></span>
<span><span class="co">#&gt;    [1,]                                            0</span></span>
<span><span class="co">#&gt;    [2,]                                            0</span></span>
<span><span class="co">#&gt;    [3,]                                            0</span></span>
<span><span class="co">#&gt;    [4,]                                            0</span></span>
<span><span class="co">#&gt;    [5,]                                            0</span></span>
<span><span class="co">#&gt;     ...                                            .</span></span>
<span><span class="co">#&gt; [9262,]                                   0.00000000</span></span>
<span><span class="co">#&gt; [9263,]                                   0.00000000</span></span>
<span><span class="co">#&gt; [9264,]                                   0.00000000</span></span>
<span><span class="co">#&gt; [9265,]                                   0.00000000</span></span>
<span><span class="co">#&gt; [9266,]                                   0.00000000</span></span>
<span><span class="co">#&gt;         sample1-92e906ae-cddb-4347-a114-bf9137761a8d</span></span>
<span><span class="co">#&gt;    [1,]                                            0</span></span>
<span><span class="co">#&gt;    [2,]                                            0</span></span>
<span><span class="co">#&gt;    [3,]                                            0</span></span>
<span><span class="co">#&gt;    [4,]                                            0</span></span>
<span><span class="co">#&gt;    [5,]                                            0</span></span>
<span><span class="co">#&gt;     ...                                            .</span></span>
<span><span class="co">#&gt; [9262,]                                   0.05273438</span></span>
<span><span class="co">#&gt; [9263,]                                   0.13085938</span></span>
<span><span class="co">#&gt; [9264,]                                   0.06835938</span></span>
<span><span class="co">#&gt; [9265,]                                   0.06054688</span></span>
<span><span class="co">#&gt; [9266,]                                   0.05664062</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;9266 x 5 SparseMatrix&gt; of type "double" [nzcount=14789 (32%)]:</span></span>
<span><span class="co">#&gt;                sample1-233e48a7-f379-4dcf-9270-958231125563 ...</span></span>
<span><span class="co">#&gt; chr1:6930410:+                                            0   .</span></span>
<span><span class="co">#&gt; chr1:6930415:+                                            0   .</span></span>
<span><span class="co">#&gt; chr1:6930419:+                                            0   .</span></span>
<span><span class="co">#&gt; chr1:6930421:+                                            0   .</span></span>
<span><span class="co">#&gt; chr1:6930428:+                                            0   .</span></span>
<span><span class="co">#&gt;            ...                                            .   .</span></span>
<span><span class="co">#&gt; chr1:6941611:-                                            0   .</span></span>
<span><span class="co">#&gt; chr1:6941614:-                                            0   .</span></span>
<span><span class="co">#&gt; chr1:6941620:-                                            0   .</span></span>
<span><span class="co">#&gt; chr1:6941622:-                                            0   .</span></span>
<span><span class="co">#&gt; chr1:6941631:-                                            0   .</span></span>
<span><span class="co">#&gt;                sample2-d03efe3b-a45b-430b-9cb6-7e5882e4faf8</span></span>
<span><span class="co">#&gt; chr1:6930410:+                                         0.02</span></span>
<span><span class="co">#&gt; chr1:6930415:+                                         0.02</span></span>
<span><span class="co">#&gt; chr1:6930419:+                                         0.02</span></span>
<span><span class="co">#&gt; chr1:6930421:+                                         0.02</span></span>
<span><span class="co">#&gt; chr1:6930428:+                                         0.02</span></span>
<span><span class="co">#&gt;            ...                                            .</span></span>
<span><span class="co">#&gt; chr1:6941611:-                                            0</span></span>
<span><span class="co">#&gt; chr1:6941614:-                                            0</span></span>
<span><span class="co">#&gt; chr1:6941620:-                                            0</span></span>
<span><span class="co">#&gt; chr1:6941622:-                                            0</span></span>
<span><span class="co">#&gt; chr1:6941631:-                                            0</span></span></code></pre></div>
<p>One advantage of the this grouping of reads per sample is that you
can easily perform per-sample operations using <code>lapply</code>
(returns a <code>list</code>) or <code>endoapply</code> (returns a
<code>DataFrame</code>):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">m</span>, <span class="va">ncol</span><span class="op">)</span></span>
<span><span class="co">#&gt; $sample1</span></span>
<span><span class="co">#&gt; [1] 3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sample2</span></span>
<span><span class="co">#&gt; [1] 2</span></span>
<span><span class="fu">endoapply</span><span class="op">(</span><span class="va">m</span>, <span class="va">ncol</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 1 row and 2 columns</span></span>
<span><span class="co">#&gt;     sample1   sample2</span></span>
<span><span class="co">#&gt;   &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; 1         3         2</span></span></code></pre></div>
<p>The <code>SparseMatrix</code> objects do not store the zeros and thus
use much less memory compared to a normal (dense) matrix. However, you
have to be careful when interpreting the values in that matrix, as they
follow a specific convention:</p>
<p><strong>Important: </strong> The zeros correspond to unobserved
read/position combinations, while all values that are ‘implicitly’
called (with a modifictation probability of less than 5%) are
represented with a value of 0.02.</p>
<p>That means if you would be manually calculating for example the
average modification probability at a given position, taking the row
mean would not be correct, as it would include the zero values that
correspond to unobserved data:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># modification probabilities at position chr1:6928850:-</span></span>
<span><span class="va">m</span><span class="op">[</span><span class="st">"chr1:6928850:-"</span>, <span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 1 row and 2 columns</span></span>
<span><span class="co">#&gt;                                  sample1        sample2</span></span>
<span><span class="co">#&gt;                           &lt;SparseMatrix&gt; &lt;SparseMatrix&gt;</span></span>
<span><span class="co">#&gt; chr1:6928850:- 0.623046875:0.099609375:0            0:0</span></span>
<span></span>
<span><span class="co"># WRONG: take the mean of all values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span><span class="st">"chr1:6928850:-"</span>, <span class="op">]</span>, <span class="va">rowMeans</span><span class="op">)</span></span>
<span><span class="co">#&gt; $sample1</span></span>
<span><span class="co">#&gt; [1] 0.2408854</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sample2</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span></span>
<span><span class="co"># CORRECT: exclude the zeros</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span><span class="st">"chr1:6928850:-"</span>, <span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">non_zero</span> <span class="op">&lt;-</span> <span class="fu">SparseArray</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SparseArray/man/SparseArray-class.html" class="external-link">nzwhich</a></span><span class="op">(</span><span class="va">x</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">non_zero</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; $sample1</span></span>
<span><span class="co">#&gt; [1] 0.3613281</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sample2</span></span>
<span><span class="co">#&gt; [1] NaN</span></span></code></pre></div>
<p>This is however rarely needed, as there are convenience functions
that will exclude the unobserved (zero) values automatically for you.
For example, you can summarize the reads in each sample using
<code><a href="../reference/addReadsSummary.html">addReadsSummary()</a></code> (see next section).</p>
</div>
<div class="section level2">
<h2 id="summarize-read-level-data">Summarize read-level data<a class="anchor" aria-label="anchor" href="#summarize-read-level-data"></a>
</h2>
<p>Summarized data can be obtained from the read-level data by
calculating a per-position summary over the reads in each sample:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addReadsSummary.html">addReadsSummary</a></span><span class="op">(</span><span class="va">se</span>, keep.reads <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                              statistics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Nmod"</span>, <span class="st">"Nvalid"</span>, <span class="st">"FracMod"</span>,</span>
<span>                                             <span class="st">"Pmod"</span>, <span class="st">"AvgConf"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>As discussed above, this will automatically exclude the non-observed
(zero) values from the data when calculating the modification
probability at each position (<code>Pmod</code> assay):</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_summary</span>, <span class="st">"Pmod"</span><span class="op">)</span><span class="op">[</span><span class="st">"chr1:6928850:-"</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;   sample1   sample2 </span></span>
<span><span class="co">#&gt; 0.3613281       NaN</span></span></code></pre></div>
<p>The summary statistics to calculate are selected using the
<code>statistics</code> argument. By default,
<code><a href="../reference/addReadsSummary.html">addReadsSummary()</a></code> will count the number of modified
(<code>Nmod</code>) and total (<code>Nvalid</code>) reads at each
position and sample, and calculate the fraction of modified bases from
the two (<code>FracMod</code>).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_summary</span>, <span class="st">"Nmod"</span><span class="op">)</span><span class="op">[</span><span class="st">"chr1:6928850:-"</span>, <span class="op">]</span></span>
<span><span class="co">#&gt; sample1 sample2 </span></span>
<span><span class="co">#&gt;       1       0</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_summary</span>, <span class="st">"Nvalid"</span><span class="op">)</span><span class="op">[</span><span class="st">"chr1:6928850:-"</span>, <span class="op">]</span></span>
<span><span class="co">#&gt; sample1 sample2 </span></span>
<span><span class="co">#&gt;       2       0</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_summary</span>, <span class="st">"FracMod"</span><span class="op">)</span><span class="op">[</span><span class="st">"chr1:6928850:-"</span>, <span class="op">]</span></span>
<span><span class="co">#&gt; sample1 sample2 </span></span>
<span><span class="co">#&gt;     0.5     NaN</span></span></code></pre></div>
<p>In the above example, we in addition also calculate the average
modification probability (<code>Pmod</code>) and the average confidence
of the modification calls per position (<code>AvgConf</code>). As we
have set <code>keep.reads = TRUE</code>, we get in addition also the
read-level assay from the input object (<code>mod_prob</code>) in which
the reads are grouped by sample:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read-level data is retained in "mod_prob" assay</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">se_summary</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Nmod"     "Nvalid"   "FracMod"  "Pmod"     "AvgConf"  "mod_prob"</span></span>
<span></span>
<span><span class="co"># ... which groups the reads by sample</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_summary</span>, <span class="st">"mod_prob"</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 9266 rows and 2 columns</span></span>
<span><span class="co">#&gt;                        sample1        sample2</span></span>
<span><span class="co">#&gt;                 &lt;SparseMatrix&gt; &lt;SparseMatrix&gt;</span></span>
<span><span class="co">#&gt; chr1:6930410:+           0:0:0         0:0.02</span></span>
<span><span class="co">#&gt; chr1:6930415:+           0:0:0         0:0.02</span></span>
<span><span class="co">#&gt; chr1:6930419:+           0:0:0         0:0.02</span></span>
<span><span class="co">#&gt; chr1:6930421:+           0:0:0         0:0.02</span></span>
<span><span class="co">#&gt; chr1:6930428:+           0:0:0         0:0.02</span></span>
<span><span class="co">#&gt; ...                        ...            ...</span></span>
<span><span class="co">#&gt; chr1:6941611:- 0:0:0.052734375            0:0</span></span>
<span><span class="co">#&gt; chr1:6941614:- 0:0:0.130859375            0:0</span></span>
<span><span class="co">#&gt; chr1:6941620:- 0:0:0.068359375            0:0</span></span>
<span><span class="co">#&gt; chr1:6941622:- 0:0:0.060546875            0:0</span></span>
<span><span class="co">#&gt; chr1:6941631:- 0:0:0.056640625            0:0</span></span>
<span></span>
<span><span class="co"># the dimensions of  read-level `se` and summarized `se_summary` are identical</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 9266    2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">se_summary</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 9266    2</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="plot-data">Plot data<a class="anchor" aria-label="anchor" href="#plot-data"></a>
</h2>
<p>The read-level data can then be visualized just like the
summary-level data using <code>plotRegion</code>, using the
<code>tracks.reads</code> argument instead of
<code>tracks.summary</code> to select the type of plot(s).</p>
<p>For reference, here we plot the summary-level data:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegion.html">plotRegion</a></span><span class="op">(</span><span class="va">se_summary</span>, region <span class="op">=</span> <span class="st">"chr1:6932700-6932800"</span>,</span>
<span>           tracks.summary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>FracMod <span class="op">=</span> <span class="st">"Point"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="read-level-data_files/figure-html/plot-summary-1.png" alt="A summary-level point plot with the fraction modified bases per position." width="576"></p>
<p>… and here we plot the read-level data of the same region:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegion.html">plotRegion</a></span><span class="op">(</span><span class="va">se</span>, region <span class="op">=</span> <span class="st">"chr1:6932700-6932800"</span>,</span>
<span>           tracks.summary <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           tracks.reads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mod_prob <span class="op">=</span> <span class="st">"Heatmap"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="read-level-data_files/figure-html/plot-region1-1.png" alt="A read-level heatmap showing modification probabilities of individual modified bases of each read grouped by sample." width="768"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegion.html">plotRegion</a></span><span class="op">(</span><span class="va">se</span>, region <span class="op">=</span> <span class="st">"chr1:6932700-6932800"</span>,</span>
<span>           tracks.summary <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           tracks.reads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mod_prob <span class="op">=</span> <span class="st">"Lollipop"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="read-level-data_files/figure-html/plot-region2-1.png" alt="A read-level point (lollipop) plot showing modification probabilities of individual modified bases of each read grouped by sample." width="768"></p>
<p>The x-axis in these plots is in “base-space”, meaning that it shows
the coordinates of genomic bases on which modifications can be
irregularly spaced. Alternatively, we can also generate these plot in
“modbase-space”, in which only modified bases are shown and the gaps
between them are removed:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegion.html">plotRegion</a></span><span class="op">(</span><span class="va">se</span>, region <span class="op">=</span> <span class="st">"chr1:6932700-6932800"</span>,</span>
<span>           tracks.summary <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           tracks.reads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mod_prob <span class="op">=</span> <span class="st">"Heatmap"</span><span class="op">)</span>,</span>
<span>           modbaseSpace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="read-level-data_files/figure-html/plot-region-modbasespace1-1.png" alt="A read-level heatmap showing modification probabilities of individual modified bases only, without interviening non-modified bases." width="768"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegion.html">plotRegion</a></span><span class="op">(</span><span class="va">se</span>, region <span class="op">=</span> <span class="st">"chr1:6932700-6932800"</span>,</span>
<span>           tracks.summary <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           tracks.reads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mod_prob <span class="op">=</span> <span class="st">"Lollipop"</span><span class="op">)</span>,</span>
<span>           modbaseSpace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="read-level-data_files/figure-html/plot-region-modbasespace2-1.png" alt="A read-level point (lollipop) plot showing modification probabilities of individual modified bases only, without interviening non-modified bases." width="768"></p>
<p>As mentioned above, the read-level data is still contained in the
summary object <code>se_summary</code>, so we can plot both
summary-level and read-level data simultaneously with this object as
input:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegion.html">plotRegion</a></span><span class="op">(</span><span class="va">se_summary</span>, region <span class="op">=</span> <span class="st">"chr1:6932700-6932800"</span>,</span>
<span>           tracks.summary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>FracMod <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Smooth"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           tracks.reads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mod_prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heatmap"</span>, <span class="st">"Lollipop"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="read-level-data_files/figure-html/plot-summary-and-read-level-1.png" alt="A plot of modification data combining different track types." width="768"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.6.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SummarizedExperiment_1.35.1 Biobase_2.65.0             </span></span>
<span><span class="co">#&gt;  [3] GenomicRanges_1.57.1        GenomeInfoDb_1.41.1        </span></span>
<span><span class="co">#&gt;  [5] IRanges_2.39.2              S4Vectors_0.43.2           </span></span>
<span><span class="co">#&gt;  [7] BiocGenerics_0.51.0         MatrixGenerics_1.17.0      </span></span>
<span><span class="co">#&gt;  [9] matrixStats_1.3.0           footprintR_0.1.1           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1            viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt;  [3] dplyr_1.1.4                 farver_2.1.2               </span></span>
<span><span class="co">#&gt;  [5] Biostrings_2.73.1           bitops_1.0-8               </span></span>
<span><span class="co">#&gt;  [7] fastmap_1.2.0               SingleCellExperiment_1.27.2</span></span>
<span><span class="co">#&gt;  [9] RCurl_1.98-1.16             GenomicAlignments_1.41.0   </span></span>
<span><span class="co">#&gt; [11] XML_3.99-0.17               digest_0.6.37              </span></span>
<span><span class="co">#&gt; [13] lifecycle_1.0.4             magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [15] compiler_4.4.1              rlang_1.1.4                </span></span>
<span><span class="co">#&gt; [17] sass_0.4.9                  tools_4.4.1                </span></span>
<span><span class="co">#&gt; [19] utf8_1.2.4                  yaml_2.3.10                </span></span>
<span><span class="co">#&gt; [21] data.table_1.16.0           rtracklayer_1.65.0         </span></span>
<span><span class="co">#&gt; [23] knitr_1.48                  S4Arrays_1.5.7             </span></span>
<span><span class="co">#&gt; [25] labeling_0.4.3              curl_5.2.2                 </span></span>
<span><span class="co">#&gt; [27] DelayedArray_0.31.11        abind_1.4-5                </span></span>
<span><span class="co">#&gt; [29] BiocParallel_1.39.0         withr_3.0.1                </span></span>
<span><span class="co">#&gt; [31] purrr_1.0.2                 desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [33] grid_4.4.1                  fansi_1.0.6                </span></span>
<span><span class="co">#&gt; [35] beachmat_2.21.5             colorspace_2.1-1           </span></span>
<span><span class="co">#&gt; [37] ggplot2_3.5.1               scales_1.3.0               </span></span>
<span><span class="co">#&gt; [39] cli_3.6.3                   rmarkdown_2.28             </span></span>
<span><span class="co">#&gt; [41] crayon_1.5.3                ragg_1.3.2                 </span></span>
<span><span class="co">#&gt; [43] generics_0.1.3              httr_1.4.7                 </span></span>
<span><span class="co">#&gt; [45] rjson_0.2.22                scuttle_1.15.4             </span></span>
<span><span class="co">#&gt; [47] cachem_1.1.0                zlibbioc_1.51.1            </span></span>
<span><span class="co">#&gt; [49] parallel_4.4.1              XVector_0.45.0             </span></span>
<span><span class="co">#&gt; [51] restfulr_0.0.15             vctrs_0.6.5                </span></span>
<span><span class="co">#&gt; [53] Matrix_1.7-0                jsonlite_1.8.8             </span></span>
<span><span class="co">#&gt; [55] patchwork_1.2.0             systemfonts_1.1.0          </span></span>
<span><span class="co">#&gt; [57] jquerylib_0.1.4             tidyr_1.3.1                </span></span>
<span><span class="co">#&gt; [59] glue_1.7.0                  pkgdown_2.1.0.9000         </span></span>
<span><span class="co">#&gt; [61] codetools_0.2-20            gtable_0.3.5               </span></span>
<span><span class="co">#&gt; [63] BiocIO_1.15.2               UCSC.utils_1.1.0           </span></span>
<span><span class="co">#&gt; [65] munsell_0.5.1               tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [67] pillar_1.9.0                htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [69] GenomeInfoDbData_1.2.12     BSgenome_1.73.0            </span></span>
<span><span class="co">#&gt; [71] R6_2.5.1                    textshaping_0.4.0          </span></span>
<span><span class="co">#&gt; [73] evaluate_0.24.0             lattice_0.22-6             </span></span>
<span><span class="co">#&gt; [75] highr_0.11                  Rsamtools_2.21.1           </span></span>
<span><span class="co">#&gt; [77] bslib_0.8.0                 Rcpp_1.0.13                </span></span>
<span><span class="co">#&gt; [79] SparseArray_1.5.31          xfun_0.47                  </span></span>
<span><span class="co">#&gt; [81] fs_1.6.4                    pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Stadler, Charlotte Soneson, Panagiotis Papasaikas, Sebastien Smallwood.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
