<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Using `footprintR` • footprintR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using `footprintR`">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">footprintR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/read-level-data.html">Exploring read-level footprinting data with `footprintR`</a></li>
    <li><a class="dropdown-item" href="../articles/using-footprintR.html">Using `footprintR`</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/fmicompbio/footprintR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Using `footprintR`</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/fmicompbio/footprintR/blob/devel/vignettes/using-footprintR.Rmd" class="external-link"><code>vignettes/using-footprintR.Rmd</code></a></small>
      <div class="d-none name"><code>using-footprintR.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><code>footprintR</code> provides tools for working with
single-molecule footprinting data in R. These include functions for
reading collapsed or read-level data from files generated by <a href="https://github.com/nanoporetech/dorado" class="external-link">Dorado</a> or <a href="https://nanoporetech.github.io/modkit" class="external-link">modkit</a> and
representation of such data as R objects, functions to manipulate such
objects and for visualization.</p>
<p>Current contributors include:</p>
<ul>
<li><a href="https://github.com/ppapasaikas" class="external-link">Panagiotis
Papasaikas</a></li>
<li><a href="https://github.com/smalseba" class="external-link">Sebastien Smallwood</a></li>
<li><a href="https://github.com/csoneson" class="external-link">Charlotte Soneson</a></li>
<li><a href="https://github.com/mbstadler" class="external-link">Michael Stadler</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>footprintR</code> can be installed from GitHub via the
<code>BiocManager</code> package:</p>
<pre><code><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"fmicompbio/footprintR"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="functionality">Functionality<a class="anchor" aria-label="anchor" href="#functionality"></a>
</h2>
<p>Here is a minimal example for using <code>footprintR</code> that
illustrates its functionality using summary-level data.</p>
<p>We start by loading the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmicompbio/footprintR" class="external-link">footprintR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="read-summary-data-from-bed-files">Read summary data from bed files<a class="anchor" aria-label="anchor" href="#read-summary-data-from-bed-files"></a>
</h3>
<p>Here, we will demonstrate how to read and visualize summary-level
data (where individual reads have been combined per genomic position).
For an example of how to work with read-level data, see the
<code><a href="../articles/read-level-data.html">vignette("read-level-data")</a></code>.</p>
<p>The example is using small example data provided as part of the
package. Here we get the file names for per-position summary
modification data in <code>bedMethyl</code> format:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># collapsed 5mC data for a small genomic window in bedMethyl format</span></span>
<span><span class="va">bedmethylfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>                              <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"modkit_pileup_1.bed.gz"</span>,</span>
<span>                                <span class="st">"modkit_pileup_2.bed.gz"</span><span class="op">)</span>,</span>
<span>                              package <span class="op">=</span> <span class="st">"footprintR"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># file with sequence of the reference genome in fasta format</span></span>
<span><span class="va">reffile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"reference.fa.gz"</span>, package <span class="op">=</span> <span class="st">"footprintR"</span><span class="op">)</span></span></code></pre></div>
<p>This allows us to read the data using
<code><a href="../reference/readBedMethyl.html">readBedMethyl()</a></code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bedmethylfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readBedMethyl.html">readBedMethyl</a></span><span class="op">(</span><span class="va">bedmethylfiles</span>,</span>
<span>                    sequence.context <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                    sequence.reference <span class="op">=</span> <span class="va">reffile</span><span class="op">)</span></span>
<span><span class="va">se</span></span>
<span><span class="co">#&gt; class: RangedSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 12020 2 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): Nmod Nvalid</span></span>
<span><span class="co">#&gt; rownames: NULL</span></span>
<span><span class="co">#&gt; rowData names(1): sequence.context</span></span>
<span><span class="co">#&gt; colnames(2): sample1 sample2</span></span>
<span><span class="co">#&gt; colData names(0):</span></span></code></pre></div>
<p>These <code>bedMethyl</code> files contain summarized data and are
read into a <code>RangedSummarizedExperiment</code> object. The rows
here correspond to genomic positions:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; UnstitchedGPos object with 12020 positions and 1 metadata column:</span></span>
<span><span class="co">#&gt;           seqnames       pos strand | sequence.context</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt; |   &lt;DNAStringSet&gt;</span></span>
<span><span class="co">#&gt;       [1]     chr1   6937686      + |              GCA</span></span>
<span><span class="co">#&gt;       [2]     chr1   6937688      + |              ACT</span></span>
<span><span class="co">#&gt;       [3]     chr1   6937689      + |              CTT</span></span>
<span><span class="co">#&gt;       [4]     chr1   6937691      + |              TAA</span></span>
<span><span class="co">#&gt;       [5]     chr1   6937696      + |              GCA</span></span>
<span><span class="co">#&gt;       ...      ...       ...    ... .              ...</span></span>
<span><span class="co">#&gt;   [12016]     chr1   6957049      - |              TGG</span></span>
<span><span class="co">#&gt;   [12017]     chr1   6957051      - |              CCT</span></span>
<span><span class="co">#&gt;   [12018]     chr1   6957052      - |              TCC</span></span>
<span><span class="co">#&gt;   [12019]     chr1   6957056      - |              CCT</span></span>
<span><span class="co">#&gt;   [12020]     chr1   6957057      - |              CCC</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>… and the columns correspond to the different samples (here
corresponding to the names of the two input files in
<code>bedmethylfiles</code>):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "sample1" "sample2"</span></span></code></pre></div>
<p>The data is stored in two matrices (assays) called <code>Nmod</code>
(the number of modified bases per sample and position) and
<code>Nvalid</code> (the number of valid total read covering that
position in each sample):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Nmod"   "Nvalid"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"Nmod"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;      sample1 sample2</span></span>
<span><span class="co">#&gt; [1,]       1       5</span></span>
<span><span class="co">#&gt; [2,]       1       0</span></span>
<span><span class="co">#&gt; [3,]       0       0</span></span>
<span><span class="co">#&gt; [4,]       0       0</span></span>
<span><span class="co">#&gt; [5,]       3       3</span></span>
<span><span class="co">#&gt; [6,]       0       0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"Nvalid"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;      sample1 sample2</span></span>
<span><span class="co">#&gt; [1,]      16      15</span></span>
<span><span class="co">#&gt; [2,]      16      15</span></span>
<span><span class="co">#&gt; [3,]       0       1</span></span>
<span><span class="co">#&gt; [4,]       1       0</span></span>
<span><span class="co">#&gt; [5,]      16      14</span></span>
<span><span class="co">#&gt; [6,]       1       0</span></span></code></pre></div>
<p>From these, you can easily calculate the fraction of
modifications:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fraction_modified</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"Nmod"</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"Nvalid"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fraction_modified</span><span class="op">)</span></span>
<span><span class="co">#&gt;      sample1   sample2</span></span>
<span><span class="co">#&gt; [1,]  0.0625 0.3333333</span></span>
<span><span class="co">#&gt; [2,]  0.0625 0.0000000</span></span>
<span><span class="co">#&gt; [3,]     NaN 0.0000000</span></span>
<span><span class="co">#&gt; [4,]  0.0000       NaN</span></span>
<span><span class="co">#&gt; [5,]  0.1875 0.2142857</span></span>
<span><span class="co">#&gt; [6,]  0.0000       NaN</span></span></code></pre></div>
<p>Please note that non-finite fractions result from positions that had
zero coverage in a given sample.</p>
</div>
<div class="section level3">
<h3 id="plot-data">Plot data<a class="anchor" aria-label="anchor" href="#plot-data"></a>
</h3>
<p>Use the <code><a href="../reference/plotRegion.html">plotRegion()</a></code> function to visualize data for a
specific region, here visualizing the accessibility using GpC
methylation (sequence context GCH), and adding a smooth line:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegion.html">plotRegion</a></span><span class="op">(</span><span class="va">se</span>,</span>
<span>           region <span class="op">=</span> <span class="st">"chr1:6939000-6946000"</span>,</span>
<span>           tracks.summary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>FracMod <span class="op">=</span> <span class="st">"PointSmooth"</span><span class="op">)</span>,</span>
<span>           sequence.context <span class="op">=</span> <span class="st">"GCH"</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-footprintR_files/figure-html/plot-region-GCH-1.png" alt="A point plot showing the fraction of modified bases per position in GCH contexts, coloured by sample." width="576"></p>
<p>Or the endogenous DNA methylation (sequence context HCG):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegion.html">plotRegion</a></span><span class="op">(</span><span class="va">se</span>,</span>
<span>           region <span class="op">=</span> <span class="st">"chr1:6939000-6946000"</span>,</span>
<span>           tracks.summary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>FracMod <span class="op">=</span> <span class="st">"PointSmooth"</span><span class="op">)</span>,</span>
<span>           sequence.context <span class="op">=</span> <span class="st">"HCG"</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-footprintR_files/figure-html/plot-region-HCG-1.png" alt="A point plot showing the fraction of modified bases per position in HCG contexts, coloured by sample." width="576"></p>
<p>Or just plotting the coverage (<code>Nvalid</code>) as a smooth
line:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegion.html">plotRegion</a></span><span class="op">(</span><span class="va">se</span>,</span>
<span>           region <span class="op">=</span> <span class="st">"chr1:6939000-6946000"</span>,</span>
<span>           tracks.summary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Nvalid <span class="op">=</span> <span class="st">"Smooth"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-footprintR_files/figure-html/plot-region-Nvalid-1.png" alt="A smoothed line plot showing the read coverage (Nvalid) per position, coloured by sample." width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.6.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SummarizedExperiment_1.35.1 Biobase_2.65.0             </span></span>
<span><span class="co">#&gt;  [3] GenomicRanges_1.57.1        GenomeInfoDb_1.41.1        </span></span>
<span><span class="co">#&gt;  [5] IRanges_2.39.2              S4Vectors_0.43.2           </span></span>
<span><span class="co">#&gt;  [7] BiocGenerics_0.51.0         MatrixGenerics_1.17.0      </span></span>
<span><span class="co">#&gt;  [9] matrixStats_1.3.0           footprintR_0.1.1           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1            farver_2.1.2               </span></span>
<span><span class="co">#&gt;  [3] dplyr_1.1.4                 R.utils_2.12.3             </span></span>
<span><span class="co">#&gt;  [5] Biostrings_2.73.1           bitops_1.0-8               </span></span>
<span><span class="co">#&gt;  [7] fastmap_1.2.0               SingleCellExperiment_1.27.2</span></span>
<span><span class="co">#&gt;  [9] RCurl_1.98-1.16             GenomicAlignments_1.41.0   </span></span>
<span><span class="co">#&gt; [11] XML_3.99-0.17               digest_0.6.37              </span></span>
<span><span class="co">#&gt; [13] lifecycle_1.0.4             magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [15] compiler_4.4.1              rlang_1.1.4                </span></span>
<span><span class="co">#&gt; [17] sass_0.4.9                  tools_4.4.1                </span></span>
<span><span class="co">#&gt; [19] utf8_1.2.4                  yaml_2.3.10                </span></span>
<span><span class="co">#&gt; [21] data.table_1.16.0           rtracklayer_1.65.0         </span></span>
<span><span class="co">#&gt; [23] knitr_1.48                  labeling_0.4.3             </span></span>
<span><span class="co">#&gt; [25] S4Arrays_1.5.7              curl_5.2.2                 </span></span>
<span><span class="co">#&gt; [27] DelayedArray_0.31.11        abind_1.4-5                </span></span>
<span><span class="co">#&gt; [29] BiocParallel_1.39.0         withr_3.0.1                </span></span>
<span><span class="co">#&gt; [31] purrr_1.0.2                 desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [33] R.oo_1.26.0                 grid_4.4.1                 </span></span>
<span><span class="co">#&gt; [35] fansi_1.0.6                 beachmat_2.21.5            </span></span>
<span><span class="co">#&gt; [37] colorspace_2.1-1            ggplot2_3.5.1              </span></span>
<span><span class="co">#&gt; [39] scales_1.3.0                cli_3.6.3                  </span></span>
<span><span class="co">#&gt; [41] rmarkdown_2.28              crayon_1.5.3               </span></span>
<span><span class="co">#&gt; [43] ragg_1.3.2                  generics_0.1.3             </span></span>
<span><span class="co">#&gt; [45] httr_1.4.7                  rjson_0.2.22               </span></span>
<span><span class="co">#&gt; [47] scuttle_1.15.4              cachem_1.1.0               </span></span>
<span><span class="co">#&gt; [49] zlibbioc_1.51.1             parallel_4.4.1             </span></span>
<span><span class="co">#&gt; [51] XVector_0.45.0              restfulr_0.0.15            </span></span>
<span><span class="co">#&gt; [53] vctrs_0.6.5                 Matrix_1.7-0               </span></span>
<span><span class="co">#&gt; [55] jsonlite_1.8.8              patchwork_1.2.0            </span></span>
<span><span class="co">#&gt; [57] systemfonts_1.1.0           jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [59] tidyr_1.3.1                 glue_1.7.0                 </span></span>
<span><span class="co">#&gt; [61] pkgdown_2.1.0.9000          codetools_0.2-20           </span></span>
<span><span class="co">#&gt; [63] gtable_0.3.5                BiocIO_1.15.2              </span></span>
<span><span class="co">#&gt; [65] UCSC.utils_1.1.0            munsell_0.5.1              </span></span>
<span><span class="co">#&gt; [67] tibble_3.2.1                pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [69] htmltools_0.5.8.1           GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">#&gt; [71] BSgenome_1.73.0             R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [73] textshaping_0.4.0           evaluate_0.24.0            </span></span>
<span><span class="co">#&gt; [75] lattice_0.22-6              highr_0.11                 </span></span>
<span><span class="co">#&gt; [77] R.methodsS3_1.8.2           Rsamtools_2.21.1           </span></span>
<span><span class="co">#&gt; [79] bslib_0.8.0                 Rcpp_1.0.13                </span></span>
<span><span class="co">#&gt; [81] SparseArray_1.5.31          xfun_0.47                  </span></span>
<span><span class="co">#&gt; [83] fs_1.6.4                    pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Stadler, Charlotte Soneson, Panagiotis Papasaikas, Sebastien Smallwood.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
